inc/elasticsearch_packages/namespace.php

```diff
@@ -0,0 +1,342 @@
+use Exception;
+
+/**
+ * Hook into search module package support.
+ *
+ * @return void
+ */
+function setup() {
+	// Check the current environment and bail if it isn't AWS.
+	if ( ! in_array( Altis\get_environment_architecture(), [ 'ec2', 'ecs' ], true ) ) {
+		return;
+	}
+
+	add_filter( 'altis.search.packages_dir', __NAMESPACE__ . '\\packages_dir', 9 );
+	add_filter( 'altis.search.create_package_id', __NAMESPACE__ . '\\create_package_id', 10, 4 );
+	add_filter( 'altis.search.get_package_id', __NAMESPACE__ . '\\get_package_id', 10, 2 );
+	add_action( 'altis.search.check_package_status', __NAMESPACE__ . '\\on_check_package_status', 10, 3 );
+	add_action( 'altis.search.dissociate_package', __NAMESPACE__ . '\\dissociate_package' );
+	add_action( 'altis.search.delete_package', __NAMESPACE__ . '\\delete_package' );
+	add_action( 'altis.search.deleted_package', __NAMESPACE__ . '\\on_deleted_package', 10, 4 );
+
+	// Remove default packages updated hook so we don't try to update indexes until
+	// packages have been associated with the domain.
+	add_action( 'plugins_loaded', function () {
+		remove_action( 'altis.search.updated_packages', 'Altis\\Enhanced_Search\\Packages\\on_updated_packages', 10, 2 );
+	}, 20 );
+}
+
+/**
+ * Set the package upload directory to an S3 directory.
+ *
+ * @param string $path The directory to store packages in.
+ * @return string
+ */
+function packages_dir( string $path ) : string {
+	if ( ! defined( 'S3_UPLOADS_BUCKET' ) || empty( S3_UPLOADS_BUCKET ) ) {
+		return $path;
+	}
+
+	return sprintf( 's3://%s/es-packages', S3_UPLOADS_BUCKET );
+}
+
+/**
+ * Get Elasticsearch Service Client.
+ *
+ * @return ElasticsearchServiceClient
+ */
+function get_es_client() : ElasticsearchServiceClient {
+	return Altis\get_aws_sdk()->createElasticsearchService( [
+		'version' => '2015-01-01',
+		'region' => Altis\get_environment_region(),
+	] );
+}
+
+/**
+ * Return the domain ID portion of the Elasticsearch Service host name.
+ */
+function get_elasticsearch_domain() : string {
+	return preg_replace( '#^search-([a-z][a-z0-9\-]+)-[a-z0-9]+\..*$#', '$1', 'search-platform-test-36c2d870-ylbu5ldvhmnb2blhnfiqxabzdq.eu-west-1.es.amazonaws.com' ); // ELASTICSEARCH_HOST );
+}
+
+/**
+ * Create the package on AES and return the package ID string.
+ *
+ * @param string|null $package_id The package file path for ES mappings.
+ * @param string $slug The package slug.
+ * @param string $file The package file path.
+ * @param bool $for_network Whether this is a network level package or not.
+ * @return string|null
+ */
+function create_package_id( ?string $package_id, string $slug, string $file, bool $for_network = false ) : ?string {
+
+	// Check file is an S3 file path.
+	if ( strpos( $file, 's3://' ) !== 0 ) {
+		// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+		trigger_error( sprintf( 'The given package filepath %s is not a valid S3 file path.', $file ), E_USER_WARNING );
```

Comment:
Sorry can't remember if I commented this before! For all these functions, can they return `WP_Error` so a message can be presented to the user, or is it not in a context where we could surface that? Typically `trigger_error` can be difficult to surface anywhere, especially for users.

Reply:
Update this function to deal with WP_Errors. It's still a bit gnarly in the UI template but functional at least. Also added to the stored error reporting.

Think realistically I'd need to add an event log as things happen over time but not sure how far to take that really - I can display the package status to end users so that might be ok from a user perspective, any errors will be in the error logs so is that sufficient? Gonna do some more testing with a live instance again tomorrow.